### PR TITLE
SYS-3965: added first stage migration of processed events mechanism to eth-bridge

### PR DESCRIPTION
**Pull request by MBrozhko34**
## Type of change/Merge

🚨What type of change is this PR? </br>
_Put an `x` in the boxes that apply_

- [ ] Release 
  - [ ] Increase versions 
  - [ ] Baseline tests passed  
  - Release type:
    - [ ] Major release 
    - [ ] Minor release 
    - [ ] Patch release 

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR._

- [ ] You describe the purpose of the PR, e.g.:
  - What does it do?
  - Highlight what important points reviewers should know about;
  - Indicates if there is something left for follow-up PRs.
- [ ] Documentation updated
- [ ] Business logic tested successfully
- [ ] Verify First, Write Last: In Substrate development, it is important that you always ensure preconditions are met and return errors at the beginning. After these checks have completed, then you may begin the function's computation.

### CODE REVIEW

File: pallets/eth-bridge/src/lib.rs [hunk 1]
```diff
@@ -731,23 +733,33 @@ pub mod pallet {
         }
     }
 
-    fn process_ethereum_event<T: Config>(event: &EthEvent) {
+    fn process_ethereum_event<T: Config>(event: &EthEvent) -> Result<(), DispatchError> {
```
Comment (thadouk) [inline pallets/eth-bridge/src/lib.rs:736]: i think we should consume any errors here, rather than bubbling up the error

File: pallets/token-manager/src/mock.rs [hunk 2]
```diff
@@ -376,6 +377,10 @@ impl ProcessedEventsChecker for TestRuntime {
     fn check_event(event_id: &EthEventId) -> bool {
         return PROCESSED_EVENTS.with(|l| l.borrow_mut().iter().any(|event| event == event_id))
     }
+    
+    fn add_event(event_id: &EthEventId, processed: bool) -> DispatchResult {
```
Comment (thadouk) [inline pallets/token-manager/src/mock.rs:381]: do we need an overwrite of the default one?

File: pallets/validators-manager/src/tests/mock.rs [hunk 3]
```diff
@@ -350,6 +351,10 @@ impl ProcessedEventsChecker for TestRuntime {
             })
         })
     }
+    
+    fn add_event(event_id: &EthEventId, processed: bool) -> DispatchResult {
```
Comment (thadouk) [inline pallets/validators-manager/src/tests/mock.rs:355]: ditto

File: pallets/token-manager/src/mock.rs (hunk 2 above)
Comment (MBrozhko34) [reply pallets/token-manager/src/mock.rs:381]: It is not possible to implement only one of the trait items without the other, this change was to resolve errors occurring as a result of adding trait functions

**Comment by MBrozhko34** [comment]
@nahuseyoum I have written some comments for the tests. So far I had created the values in the mock (although partially incorrectly as the successful test case does not pass) and identified the edge cases. The tests themselves still need to reach the required function process_ethereum_event().

**Review by ivan-cholakov — APPROVED** [review]

File: pallets/eth-bridge/src/lib.rs [hunk 4]
```diff
@@ -223,8 +223,11 @@ pub mod pallet {
                 BoundedVec<(BoundedVec<u8, TypeLimit>, BoundedVec<u8, ValueLimit>), ParamsLimit>,
             caller_id: BoundedVec<u8, CallerIdLimit>,
         },
-        /// EventProcessed(bool, EthEventId)
-        EventProcessed {
+        EventProcessingAccepted {
+            accepted: bool,
```
Comment (nahuseyoum) [inline pallets/eth-bridge/src/lib.rs:227]: I think we can remove the boolean because the name suggest the outcome

File: pallets/eth-bridge/src/lib.rs [hunk 5]
```diff
@@ -223,8 +223,11 @@ pub mod pallet {
                 BoundedVec<(BoundedVec<u8, TypeLimit>, BoundedVec<u8, ValueLimit>), ParamsLimit>,
             caller_id: BoundedVec<u8, CallerIdLimit>,
         },
-        /// EventProcessed(bool, EthEventId)
-        EventProcessed {
+        EventProcessingAccepted {
```
Comment (nahuseyoum) [inline pallets/eth-bridge/src/lib.rs:226]: I would also re-use the same name as EthereumEvents: `EventAccepted` and `EventRejected`.

File: pallets/eth-bridge/src/lib.rs [hunk 6]
```diff
@@ -824,28 +830,30 @@ pub mod pallet {
...
+        if T::ProcessedEventsChecker::processed_event_exists(&event.event_id.clone()) {
+            log::error!("💔 Event already processed, duplicate event");
 
-        <Pallet<T>>::deposit_event(Event::<T>::EventProcessed {
-            accepted: event_accepted,
-            eth_event_id: event.event_id.clone(),
-        });
-        // Add record of succesful processing via ProcessedEventsChecker
-        T::ProcessedEventsChecker::add_processed_event(&event.event_id.clone(), event_accepted);
+            <Pallet<T>>::deposit_event(Event::<T>::DuplicateEventSubmission {
```
Comment (nahuseyoum) [inline pallets/eth-bridge/src/lib.rs:839]: I see why you are emitting an event here, I am assuming its because we don't want to error here which is correct. But in this particular case, the event we are processing has already been processed so Its "safe" to return an error and undo any storage changes related to processing this event.

File: pallets/eth-bridge/src/tests/mock.rs [hunk 7]
```diff
@@ -181,8 +184,28 @@ pub fn create_confirming_author(author_id: u64) -> Author<TestRuntime> {
...
+        .unwrap();
+    EthereumEventsPartition::new(
+        EthBlockRange { start_block: 1, length: 1000 },
+        0,
+        false,
+        partition,
+    )
+}
+
+pub fn create_second_mock_event_partition(events: EthEvent) -> EthereumEventsPartition {
```
Comment (nahuseyoum) [inline pallets/eth-bridge/src/tests/mock.rs:198]: You can update `create_mock_event_partition` to take an extra argument. This avoids creating a second function.

**Review by nahuseyoum — APPROVED** [review]

File: pallets/avn/src/lib.rs [hunk 8]
```diff
@@ -703,13 +703,17 @@ impl<ValidatorId: Member> Enforcer<ValidatorId> for () {
 }
 
 pub trait ProcessedEventsChecker {
-    fn check_event(event_id: &EthEventId) -> bool;
+    fn processed_event_exists(event_id: &EthEventId) -> bool;
+
```
Comment (thadouk) [inline pallets/avn/src/lib.rs:707]:
```suggestion
```

File: pallets/eth-bridge/src/lib.rs [hunk 9]
```diff
@@ -219,9 +223,13 @@ pub mod pallet {
...
-        /// EventProcessed(bool, EthEventId)
-        EventProcessed {
-            accepted: bool,
+        EventAccepted {
+            eth_event_id: EthEventId,
+        },
+        EventRejected {
+            eth_event_id: EthEventId,
+        },
+        DuplicateEventSubmission {
```
Comment (thadouk) [inline pallets/eth-bridge/src/lib.rs:232]: this is not used.

File: pallets/eth-bridge/src/lib.rs [hunk 10]
```diff
@@ -296,13 +304,16 @@ pub mod pallet {
     pub enum Error<T> {
         CorroborateCallFailed,
         DuplicateConfirmation,
+        DuplicateEventSubmission,
```
Comment (thadouk) [inline pallets/eth-bridge/src/lib.rs:307]: this is not used

File: pallets/eth-bridge/src/lib.rs [hunk 11]
```diff
@@ -219,9 +223,13 @@ pub mod pallet {
...
         },
-        /// EventProcessed(bool, EthEventId)
-        EventProcessed {
-            accepted: bool,
+        EventAccepted {
+            eth_event_id: EthEventId,
+        },
+        EventRejected {
+            eth_event_id: EthEventId,
+        },
```
Comment (thadouk) [inline pallets/eth-bridge/src/lib.rs:231]: why use two events for signaling the processing events?

File: pallets/eth-bridge/src/lib.rs [hunk 12]
```diff
@@ -794,48 +806,62 @@ pub mod pallet {
     fn process_ethereum_events_partition<T: Config>(
         active_range: &ActiveEthRange,
         partition: &EthereumEventsPartition,
-    ) {
+    ) -> Result<(), DispatchError> {
```
Comment (thadouk) [inline pallets/eth-bridge/src/lib.rs:809]: This function should be infallible.
Once a voting round is concluded, the processing of each event should not affect the processing of the batch.

The errors should be handled, the appropriate events emited and processing continues.
Otherwise, the chain will be stuck in a loop of voting for the last vote and rejecting the processing.

File: pallets/eth-bridge/src/lib.rs [hunk 13]
```diff
@@ -794,48 +806,62 @@ pub mod pallet {
...
         }
+
+        Ok(())
     }
 
-    fn process_ethereum_event<T: Config>(event: &EthEvent) {
-        // TODO before processing ensure that the event has not already been processed
-        match T::BridgeInterfaceNotification::on_event_processed(&event) {
+    fn process_ethereum_event<T: Config>(event: &EthEvent) -> Result<(), DispatchError> {
+        ensure!(
```
Comment (thadouk) [inline pallets/eth-bridge/src/lib.rs:839]: we shouldn't error here. If the event has been processed, we avoid double processing, signal/log/emit the appropriate info and return.

**Review by thadouk — CHANGES_REQUESTED** [review]

File: pallets/eth-bridge/src/lib.rs (hunk 11 above)
Comment (MBrozhko34) [reply pallets/eth-bridge/src/lib.rs:231]: Would it not be due to the fact that there are two possibilities in the event being accepted vs rejected and to distinguish between these two circumstances?

File: pallets/eth-bridge/src/lib.rs (hunk 13 above)
Comment (nahuseyoum) [reply pallets/eth-bridge/src/lib.rs:839]: If we call this function with an event that has already been processed, it's because the upstream functions did not validate properly.

File: pallets/eth-bridge/src/lib.rs (hunk 11 above)
Comment (nahuseyoum) [reply pallets/eth-bridge/src/lib.rs:231]: We have connected components that are currently looking for these events and its a lesser change to maintain the setup. Its also easier to search and filter instead of using event args.

File: pallets/eth-bridge/src/lib.rs (hunk 13 above)
Comment (thadouk) [reply pallets/eth-bridge/src/lib.rs:839]: if the authors/collators have reached consensus on the tier1 state then we shouldn't block it.
There are valid cases that this can occur:
1. events added and being processed while upgrading.
2. events added to tier2 due starting from an earlier last block
We should definitely mark the case, but not block the event processing mechanism.
The error will cause the last vote to be stuck and recasted, only to be rejected again and again in a loop.

Comment (nahuseyoum) [reply pallets/eth-bridge/src/lib.rs:839]: In both of these scenarios, and possibly others, I would expect the collators to check first if the event is valid before submitting it (ex: already processed or `0x000`..). 
But in any case, this function is too low level to be required to be infallible. I think we can still throw an error here and catch it on `process_ethereum_events_partition` and log it there and move on to the next event? 
It feels like making an distinction between a bad event (one that is already processed, or is 0 
 - something we should add in the validation i guess :smile: ) and a valid event that fails to be processed makes sense which is what this low level function is doing.

Comment (MBrozhko34) [reply pallets/eth-bridge/src/lib.rs]: I am happy with either solution, although perhaps handling the error within the `process_ethereum_events_partition` function is the best solution.

Comment (MBrozhko34) [reply pallets/eth-bridge/src/lib.rs:845]: I have pushed a [commit](https://github.com/AventusProtocolFoundation/avn-parachain/pull/390/commits/4b74809faf44e3f2181cea8496b8180eab997bb4) the changes @nahuseyoum suggested @thadouk . This can be reverted if necessary although I believe it should be ok

File: pallets/eth-bridge/src/lib.rs [hunk 14]
```diff
@@ -801,15 +813,25 @@ pub mod pallet {
...
-                        process_ethereum_event::<T>(&discovered_event.event);
+                        match process_ethereum_event::<T>(&discovered_event.event) {
+                            Ok(_) => {},
+                            Err(_) => {
+                                log::error!("💔 Duplicate Event Submission");
+                                <Pallet<T>>::deposit_event(Event::<T>::DuplicateEventSubmission {
+                                    eth_event_id: discovered_event.event.event_id.clone(),
+                                });
+                            },
+                        }
```
Comment (thadouk) [inline pallets/eth-bridge/src/lib.rs:824]:
```suggestion
                        if process_ethereum_event::<T>(&discovered_event.event).is_err() {
                                log::error!("💔 Duplicate Event Submission");
                                <Pallet<T>>::deposit_event(Event::<T>::DuplicateEventSubmission {
                                    eth_event_id: discovered_event.event.event_id.clone(),
                                });
                        }
```

**Review by thadouk — APPROVED** [review]

File: pallets/eth-bridge/src/lib.rs (hunk 14 above)
Comment (nahuseyoum) [reply pallets/eth-bridge/src/lib.rs:824]: Maybe we should have a generic `InvalidEventSubmission` instead of `DuplicateEventSubmission`? We can then use this for 0 transaction hashes too (not just existing events)
